### PR TITLE
[SPARK-41437][SQL] Do not optimize the input query twice for v1 write fallback

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

For v2 DML commands, the input query is a child plan and will be optimized by the optimizer. In the execution path, if the v2 table supports v1 fallback, we will create a DataFrame from the optimized input query plan, and pass the DataFrame to v1 write API.

This means that the input query gets optimized twice: once during the DML command planning, once with the extra DataFrame creation. This is a waste of optimizing time and can cause issues if some custom optimizer rules are not idempotent.

This PR adds a new field `analyzedQuery: Option[LogicalPlan]` to the v2 commands that support v1 fallback. This field will be filled at the end of analysis and will be used to create that v1 fallback DataFrame.

### Why are the changes needed?

avoid double optimizing the input query plan of v2 DML commands

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new test

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/connector/V1WriteFallbackSuite.scala [hunk 1]
```diff
@@ -132,17 +132,21 @@ class V1WriteFallbackSuite extends QueryTest with SharedSparkSession with Before
     assert(e3.getMessage.contains("schema"))
   }
 
-  test("fallback writes should only analyze plan once") {
+  test("fallback writes should only analyze/optimize plan once") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/V1WriteFallbackSuite.scala:135]:
```suggestion
  test("SPARK-41437: fallback writes should only analyze/optimize plan once") {
```

**Comment by cloud-fan** [comment]
cc @viirya @gengliangwang

File: sql/core/src/test/scala/org/apache/spark/sql/connector/V1WriteFallbackSuite.scala [hunk 2]
```diff
@@ -435,3 +439,16 @@ object OnlyOnceRule extends Rule[LogicalPlan] {
 
   }
 }
+
+// A rule that fails if the input query of a V2WriteCommand is optimized twice
+object OnlyOnceOptimizerRule extends Rule[LogicalPlan] {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/V1WriteFallbackSuite.scala:444]: Thank you for adding this.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
BTW, are we going to backport this, @cloud-fan ? It seems that the JIRA is filed as 'Bug' for `3.4.0` because of the idempotency issue.

**Comment by cloud-fan** [comment]
This changes the plan signature so I'm a bit hesitant to backport. Besides, people can fix issues by themselves, if they can make their custom optimizer rules idempotent. I think master branch should be good enough.

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Review by viirya — COMMENTED** [review]
lgtm

**Pull request closed without merging**